fixes #6441 - allows filtering of parameters per controller

dustints opened the pull request:
there is an issue where certain parameters for particular
controllers are causing the log to fill up. This allows the ability
to filter parameters by controller.

This allows each controller to add additional parameters to filter.
In a controller, to filter parameters name and description, you can add the following:

include Foreman::Controller::FilterParameters
filter_parameters :name, :description

Commit:
fixes #6441 - allows filtering of parameters per controller

There is an issue where certain parameters for particular
controllers are causing the log to fill up. This allows the ability
to filter parameters by controller.

This allows each controller to add additional parameters to filter.
In a controller, to filter parameters name and description, you can add
the following:

include Foreman::Controller::FilterParameters
filter_parameters :name, :description

domcleal commented:
Administrative stuff: please note that commit messages should reference a redmine in the Foreman project, no BZ, and be a single line (you may put long text in the PR desc).

domcleal commented:
Note you have test failures still on 1.8.7.

dustints commented:
@domcleal thanks, updated with
- fix to commit message and foreman issue created
- syntax issues for 1.8.7

domcleal commented:
@witlessbird could you review please?  I'm not at all familiar with this part of the stack.

I'm a bit concerned at maintaining a partial duplicate of something that's in core Rails.

dmitri-d commented:
Any reason not to use http://api.rubyonrails.org/classes/ActionDispatch/Http/FilterParameters.html?

dustints commented:
@domcleal, @witlessbird  updated:
- using ActionDispatch::Http::ParameterFilter for filtering the parameters
- altering the event payload so we don't need to maintain dup code in rails core

dmitri-d commented:
A few more questions: 
- with this request params are going to be processed and logged twice: once by `ActionController::LogSubscriber` and once by `CustomAcLogSubscriber`; I'm not sure disabling rails logger without replicating its output would be great, as it provides more details compared to `CustomAcLogSubscriber`.
- Why did you decide to log request params yourself? Updating `request.filtered_parameters` would do the trick and won't result in redundant logging. If implemented along the lines of https://github.com/rails/rails/blob/master/actionpack/lib/action_controller/metal/params_wrapper.rb config can be kept in the controller itself...
- And lastly, why not add another field to config.filter_parameters? Are there cases when :content  field doesn't need to be filtered?

dustints commented:
@witlessbird,

thanks for looking at this for me.
- The params shouldn't be processed twice, since in the intializer (https://github.com/dustint-rh/foreman/blob/log_filter_params_by_request_path/config/initializers/custom_ac_log_subscriber.rb#L9) we're unsubscribing `ActionController::LogSubscriber` 's `start_processing` in favor for `CustomAcLogSubscriber`'s.
- I wasn't successful in trying to update `request.filtered_parameters`  in a before_filter. Even though I was able to change the values in filtered_parameters, it still logged the original requested params unchanged. 
- I wanted to filter the `content` param in context of the ContentUploadsController because it was logging a ton of hex. A lot of the other controllers take `content` too but don't seem to have the same behavior. I didn't think it was right to just filter all `content` params across the application.

dmitri-d commented:
Fair enough re: selective `content` filtering. 

I don't think you can rely on a filter callback to modify parameter logging. In https://github.com/rails/rails/blob/master/actionpack/lib/action_controller/base.rb#L241 instrumentation is handled before callbacks...

I think something like this should do the trick though: 

```
module Foreman
  module Middleware
    module CustomParameterFilter
      CONFIG = lambda{ Foreman::Middleware::CustomAcLogSubscriberConfig.config[:filter_parameters_by_request_path] }

      def def process_action(*args)
        if needs_filtering?(*args)
          request.filtered_parameters.merge!(filter_params(request.filtered_parameters, filter_from_path(*args))
        end

        super
      end   
...
```

and then `include Foreman::Middleware::CustomParameterFilter` in ApplicationController. I think this would be preferable, as we wouldn't be replacing/duplicating Rails' default logging. Also, personally I'd prefer if additional parameter filtering was defined in the controller itself (as opposed to plugin config block).

dLobatog commented on test/lib/middleware/custom_ac_log_subscriber_test.rb:
```diff
@@ -0,0 +1,47 @@
+  def test_start_processing_not_filtered_due_to_params_not_in_config
+    expect_params_not_filtered
+    @event = OpenStruct.new({:payload => {:path => '/katello/foo_path/blah', :params => {'john' => 'bar', 'snow' => 'blah', 'dont' => 'know' }}})
+    @log_subscriber.start_processing(@event)
```
@log_subscriber.start_processing(@event) could go in a teardown block, since it has to run in every test

dustints commented:
@witlessbird thanks that worked!
I changed the configuration so that filtered_parameters can be specified using a class method on the controller. Decided that the module can be included where it is needed instead of putting in the api base controller or application controller.

@eLobato I updated the tests, they've changed quite a bit.

dLobatog commented:
@dustint-rh this looks much more usable than adding the filtered param routes like I saw in the Katello PR, thanks. It's ready to merge after a rebase of the last commits, it needs to be in this format http://projects.theforeman.org/projects/foreman/wiki/Reviewing_patches-commit_message_format

Also the title is now wrong as it doesn't filter parameters by "path", right?

dustints commented:
@eLobato updated: squashed and changed the commit title and description.

domcleal commented on lib/foreman/controller/filter_parameters.rb:
```diff
@@ -0,0 +1,37 @@
+module Foreman::Controller::FilterParameters
+  def self.included(base)
+    base.extend(ClassMethods)
```
I think you can now use AS::Concern and move to app/controllers/concerns

domcleal commented:
Thanks for the refactorings, this is looking much easier to maintain.

dLobatog commented:
@dustint-rh modules for controllers should go in `app/controller/concerns/foreman/controller` , you'll need your module to to `extend ActiveSupport::Concern` and move the test to a different dir, in fact I'd recommend moving it in `test/functional/application_controller_subclass_test.rb`

dustints commented:
@eLobato thanks I updated so that it's a concern now and moved the tests into `test/functional/application_controller_subclass_test.rb`

dLobatog commented:
Merged as https://github.com/theforeman/foreman/commit/cb49ec827efed8197791b0c58c781d41a5e43f51 , thanks @dustint-rh . I removed that dangling debugger in the test file :smile:

dustints commented:
@eLobato thanks bunches!
